### PR TITLE
nixos/hatsu: init module

### PR DESCRIPTION
**Pull request by kwaa**
## Description of changes

Add module for [Hatsu](https://github.com/importantimport/hatsu) ([package](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/by-name/ha/hatsu/package.nix#L33)), Have tested on my machine using the following configuration:

```nix
{
  services.hatsu = {
    enable = true;
    settings = {
      HATSU_LISTEN_PORT = 4040;
      HATSU_DOMAIN = "hatsu.local";
      HATSU_PRIMARY_ACCOUNT = "kwaa.dev";
    };
  };
}
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by kwaa** [comment]
@NickCao Would you mind taking a look? Thanks!

File: nixos/modules/services/web-apps/hatsu.nix [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+          };
+
+          HATSU_LISTEN_HOST = lib.mkOption {
+            type = lib.types.str;
+            default = "127.0.0.1";
+            description = "Host where hatsu should listen for incoming requests.";
+          };
+
+          HATSU_LISTEN_PORT = lib.mkOption {
+            type = lib.types.port;
```
Comment (NickCao) [inline nixos/modules/services/web-apps/hatsu.nix:57]:
```suggestion
            type = lib.types.port;
            apply = toString;
```

File: nixos/modules/services/web-apps/hatsu.nix [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  options.services.hatsu = {
+    enable = lib.mkEnableOption "Self-hosted and fully-automated ActivityPub bridge for static sites";
+
+    package = lib.mkPackageOption pkgs "hatsu" { };
+
+    dataDir = lib.mkOption {
+      type = lib.types.str;
+      default = "/var/lib/hatsu";
+      description = "Data directory for hatsu.";
+    };
```
Comment (NickCao) [inline nixos/modules/services/web-apps/hatsu.nix:23]: I don't think a custom datadir (not managed by systemd) would work well with DynamicUser. We might drop this option as well.

File: nixos/modules/services/web-apps/hatsu.nix (hunk 1 above)
Comment (kwaa) [reply nixos/modules/services/web-apps/hatsu.nix:52]: Fixed, I also changed the value of `freeformType`

File: nixos/modules/services/web-apps/hatsu.nix (hunk 2 above)
Comment (kwaa) [reply nixos/modules/services/web-apps/hatsu.nix]: Fixed as `/var/lib/hatsu`

**Comment by kwaa** [comment]
I have retested on my machine and the modified module works fine.

```nix
{
  services.hatsu = {
    enable = true;
    settings = {
      HATSU_LISTEN_PORT = 4040;
      HATSU_DOMAIN = "hatsu.local";
      HATSU_PRIMARY_ACCOUNT = "kwaa.dev";
    };
  };
}
```

File: nixos/modules/services/web-apps/hatsu.nix [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        DynamicUser = true;
+        ExecStart = "${lib.getExe cfg.package}";
+        Restart = "on-failure";
+        StateDirectory = "hatsu";
+        Type = "simple";
+        WorkingDirectory = "/var/lib/hatsu";
```
Comment (NickCao) [inline nixos/modules/services/web-apps/hatsu.nix:93]:
```suggestion
        WorkingDirectory = "%S/hatsu";
```

Comment (NickCao) [reply nixos/modules/services/web-apps/hatsu.nix:93]: A final nitpick and we are good to go.

Comment (kwaa) [reply nixos/modules/services/web-apps/hatsu.nix]: Fixed

**Review by NickCao — APPROVED** [review]

**Pull request merged**